### PR TITLE
add PytatoKeyBuilder, persistent_dict test

### PR DESCRIPTION
**Pull request by matthiasdiener**
Alternative to #457. 

See #547 for a more in-depth discussion.

_Please squash_

### CODE REVIEW

File: pytato/analysis/__init__.py [hunk 1]
```diff
@@ -453,4 +454,32 @@ def get_num_call_sites(outputs: Union[Array, DictOfNamedArrays]) -> int:
...
+            PersistentHashWalkMapper(key_hash)(key)
+
+    update_for_Product = update_for_pymbolic_expression  # noqa: N815
+    update_for_Sum = update_for_pymbolic_expression  # noqa: N815
+    update_for_If = update_for_pymbolic_expression  # noqa: N815
+    update_for_LogicalOr = update_for_pymbolic_expression  # noqa: N815
+    update_for_Call = update_for_pymbolic_expression  # noqa: N815
+    update_for_Comparison = update_for_pymbolic_expression  # noqa: N815
+    update_for_Quotient = update_for_pymbolic_expression  # noqa: N815
+    update_for_Power = update_for_pymbolic_expression  # noqa: N815
```
Comment (matthiasdiener) [inline pytato/analysis/__init__.py:481]: This could probably be improved, but I'm not sure how.

File: pytato/analysis/__init__.py [hunk 2]
```diff
@@ -453,4 +454,32 @@ def get_num_call_sites(outputs: Union[Array, DictOfNamedArrays]) -> int:
...
+
+# {{{ PytatoKeyBuilder
+
+class PytatoKeyBuilder(LoopyKeyBuilder):
+    """A custom :class:`pytools.persistent_dict.KeyBuilder` subclass
+    for objects within :mod:`pytato`.
+    """
+
+    def update_for_ndarray(self, key_hash: Any, key: Any) -> None:
+        self.rec(key_hash, hash(key.data.tobytes()))  # type: ignore[no-untyped-call]
```
Comment (matthiasdiener) [inline pytato/analysis/__init__.py:466]: Should this go somewhere else (e.g., in the pytools KeyBuilder itself)?

File: pytato/analysis/__init__.py (hunk 1 above)
Comment (matthiasdiener) [reply pytato/analysis/__init__.py]: Perhaps https://github.com/inducer/pymbolic/pull/125 resolves it.

**Comment by matthiasdiener** [comment]
This is ready for review @inducer. The current version is able to hash the DAG of a 2-rank `smoke_test_ks_3d` run.

**Comment by matthiasdiener** [comment]
A gentle ping for review @inducer. This PR is helpful with e.g. #498.

File: pytato/analysis/__init__.py (hunk 1 above)
Comment (matthiasdiener) [reply pytato/analysis/__init__.py]: I'm guessing this PR is waiting for https://github.com/inducer/pymbolic/pull/125 ?

**Comment by inducer** [comment]
See also #547.

File: pytato/analysis/__init__.py (hunk 1 above)
Comment (inducer) [reply pytato/analysis/__init__.py]: Yep, https://github.com/inducer/pymbolic/pull/125 has resolved this, this bit should no longer be necessary. :grin:

File: pytato/analysis/__init__.py [hunk 3]
```diff
@@ -564,4 +565,47 @@ def get_num_call_sites(outputs: Array | DictOfNamedArrays) -> int:
...
+    update_for_LogicalOr = LoopyKeyBuilder.update_for_pymbolic_expression  # noqa: N815
+    update_for_Lookup = LoopyKeyBuilder.update_for_pymbolic_expression  # noqa: N815
+    update_for_Power = LoopyKeyBuilder.update_for_pymbolic_expression  # noqa: N815
+    update_for_Product = LoopyKeyBuilder.update_for_pymbolic_expression  # noqa: N815
+    update_for_Quotient = LoopyKeyBuilder.update_for_pymbolic_expression  # noqa: N815
+    update_for_Remainder = LoopyKeyBuilder.update_for_pymbolic_expression  # noqa: N815
+    update_for_RightShift = LoopyKeyBuilder.update_for_pymbolic_expression  # noqa: N815
+    update_for_Subscript = LoopyKeyBuilder.update_for_pymbolic_expression  # noqa: N815
+    update_for_Sum = LoopyKeyBuilder.update_for_pymbolic_expression  # noqa: N815
+    update_for_Variable = LoopyKeyBuilder.update_for_pymbolic_expression  # noqa: N815
```
Comment (inducer) [inline pytato/analysis/__init__.py:607]: All this can go now that https://github.com/inducer/pymbolic/pull/125/ is in.

Comment (matthiasdiener) [reply pytato/analysis/__init__.py]: Removed in [70a6887](https://github.com/inducer/pytato/pull/459/commits/70a688753adaba242b796b990321202a2db4f4ab)

File: pytato/analysis/__init__.py [hunk 4]
```diff
@@ -564,4 +565,47 @@ def get_num_call_sites(outputs: Array | DictOfNamedArrays) -> int:
...
+
+    def update_for_ndarray(self, key_hash: Any, key: Any) -> None:
+        self.rec(key_hash, key.data.tobytes())
+
+    def update_for_TaggableCLArray(self, key_hash: Any, key: Any) -> None:
+        self.rec(key_hash, key.get())
+
+    def update_for_Array(self, key_hash: Any, key: Any) -> None:
+        # CL Array
+        self.rec(key_hash, key.get())
```
[the remark this answers is not in the document]
Comment (matthiasdiener) [reply pytato/analysis/__init__.py]: Added in [70a6887](https://github.com/inducer/pytato/pull/459/commits/70a688753adaba242b796b990321202a2db4f4ab)

File: pytato/analysis/__init__.py (hunk 1 above)
Comment (matthiasdiener) [reply pytato/analysis/__init__.py]: Removed in [70a6887](https://github.com/inducer/pytato/pull/459/commits/70a688753adaba242b796b990321202a2db4f4ab)

**Comment by matthiasdiener** [comment]
This is ready for another review @inducer .

**Comment by inducer** [comment]
Thanks!

**Pull request merged**